presto-main/src/main/java/com/facebook/presto/execution/scheduler/LegacySqlQueryScheduler.java

```diff
@@ -459,9 +505,182 @@ else if (!result.getBlocked().isDone()) {
                 // get all sections ready for execution
                 .filter(this::isReadyForExecution)
                 .limit(maxConcurrentMaterializations - runningPlanSections)
+                .map(this::tryCostBasedOptimize)
                 .collect(toImmutableList());
     }
 
+    /**
+     * Utility function to invoke runtime cost-based optimizer, which determines if the probe and build side of a JoinNode should be swapped,
+     * based on the statistics of the temporary table holding materialized exchange outputs from finished children sections.
+     */
+    private StreamingPlanSection tryCostBasedOptimize(StreamingPlanSection section)
+    {
+        // no need to do runtime optimization if no materialized exchange data is utilized by the section.
+        if (!isRuntimeOptimizerEnabled(session) || getExchangeMaterializationStrategy(session) != ALL || section.getChildren().isEmpty()) {
```

Comment:
I am wondering  the condition`getExchangeMaterializationStrategy(session) != ALL` is necessary? -- What we really want is `tryCostBasedOptimize` only get called for non-root sections right? In that case we probably just want to sure `tryCostBasedOptimize` doesn't get called in `getSectionsReadyForExecution` for non-root section? :)

Reply:
I guess you mean non-leaf section right? Yeah, when materialization is not enabled, there will be only one section, which has no children, and so will not trigger `tryCostBasedOptimize`

Reply:
@pguofb : Right, non-leaf section :)